File: packages/core/src/contract/file-activity.ts
import type {
  FileActivityKind,
  ReferencedSessionHead,
  SessionFileActivity,
  ToolPart,
} from "./session.js";

export type FileActivityResult = SessionFileActivity & ReferencedSessionHead;

export interface FileToolOperation {
  kind: FileActivityKind;
  path: string;
}

const FILE_TOOL_KINDS: Readonly<Record<string, FileActivityKind>> = {
  read: "read",
  read_file: "read",
  read_file_v2: "read",
  read_text_file: "read",
  readfile: "read",
  view_image: "read",
  apply_patch: "edit",
  edit: "edit",
  edit_file: "edit",
  edit_file_v2: "edit",
  editfile: "edit",
  multiedit: "edit",
  notebookedit: "edit",
  patch: "edit",
  search_replace: "edit",
  str_replace: "edit",
  create_file: "write",
  write: "write",
  write_file: "write",
  writefile: "write",
  delete: "delete",
  delete_file: "delete",
};

const PATCH_KINDS: Readonly<Record<string, FileActivityKind>> = {
  edit_file: "edit",
  update_file: "edit",
  write_file: "write",
  delete_file: "delete",
  move_file: "edit",
};

function toRecord(value: unknown) {
  if (value && typeof value === "object" && !Array.isArray(value)) {
    return value as Record<string, unknown>;
  }
  return {};
}

function looksLikeFilePath(value: string) {
  const text = value.trim();
  if (!text || text.length > 300 || text.includes("\n")) return false;
  if (/^[a-z]+:\/\//i.test(text) || /[<>{}]/.test(text)) return false;
  if (text.startsWith("/") || text.startsWith("./") || text.startsWith("../")) return true;
  if (text.startsWith("~/") || text.includes("/") || text.includes("\\")) return true;
  return /^[A-Za-z0-9_.@-]+\.[A-Za-z0-9_-]+$/.test(text);
}

function isPathKey(key: string) {
  const normalized = key.trim().toLowerCase();
  if (!normalized) return false;
  if (
    normalized.includes("command") ||
    normalized.includes("content") ||
    normalized.includes("text") ||
    normalized.includes("prompt") ||
    normalized.includes("url") ||
    normalized.includes("body") ||
    normalized.includes("title") ||
    normalized.includes("description") ||
    normalized === "cwd" ||
    normalized === "workdir" ||
    normalized === "directory"
  ) {
    return false;
  }
  return normalized.includes("file") || normalized.includes("path");
}

function collectPaths(value: unknown, key: string, paths: Set<string>, depth = 0): void {
  if (value == null || depth > 4) return;
  if (typeof value === "string") {
    if (isPathKey(key) && looksLikeFilePath(value)) paths.add(value.trim());
    return;
  }
  if (Array.isArray(value)) {
    for (const item of value) collectPaths(item, key, paths, depth + 1);
    return;
  }
  if (typeof value === "object") {
    for (const [nestedKey, nested] of Object.entries(value as Record<string, unknown>)) {
      collectPaths(nested, nestedKey, paths, depth + 1);
    }
  }
}

function extractPatchOperations(input: unknown): FileToolOperation[] {
  const record = toRecord(input);
  const content = Array.isArray(input)
    ? input
    : Array.isArray(record.content)
      ? record.content
      : [];

  return content.flatMap((entry) => {
    const patch = toRecord(entry);
    const kind = typeof patch.type === "string" ? PATCH_KINDS[patch.type] : undefined;
    const pathValue = patch.path || patch.old_path;
    const path = typeof pathValue === "string" ? pathValue.trim() : "";
    return kind && path ? [{ kind, path }] : [];
  });
}

export function classifyFileTool(part: Pick<ToolPart, "tool">): FileActivityKind | null {
  return FILE_TOOL_KINDS[part.tool.trim().toLowerCase()] ?? null;
}

export function extractFileToolOperations(
  part: Pick<ToolPart, "tool" | "state">,
): FileToolOperation[] {
  const patchOperations = extractPatchOperations(part.state.input);
  if (patchOperations.length > 0) return patchOperations;

  const kind = classifyFileTool(part);
  if (!kind) return [];

  const paths = new Set<string>();
  collectPaths(part.state.input, "", paths);
  return [...paths].map((path) => ({ kind, path }));
}

File: crates/codesesh-core/src/agents/file_activity.rs
use crate::contract::{Message, MessagePart, SessionFileActivity, SessionHead};
use regex::Regex;
use serde_json::Value;
use std::{
    collections::{HashMap, HashSet},
    sync::LazyLock,
};

fn kind(tool: &str) -> Option<&'static str> {
    match tool.trim().to_lowercase().as_str() {
        "read" | "read_file" | "read_file_v2" | "read_text_file" | "readfile" | "view_image" => {
            Some("read")
        }
        "apply_patch" | "edit" | "edit_file" | "edit_file_v2" | "editfile" | "multiedit"
        | "notebookedit" | "patch" | "search_replace" | "str_replace" => Some("edit"),
        "create_file" | "write" | "write_file" | "writefile" => Some("write"),
        "delete" | "delete_file" => Some("delete"),
        _ => None,
    }
}
fn path_key(key: &str) -> bool {
    let key = key.trim().to_lowercase();
    ![
        "command",
        "content",
        "text",
        "prompt",
        "url",
        "body",
        "title",
        "description",
    ]
    .iter()
    .any(|excluded| key.contains(excluded))
        && !matches!(key.as_str(), "cwd" | "workdir" | "directory")
        && (key.contains("file") || key.contains("path"))
}
fn file_path(text: &str) -> bool {
    let text = text.trim();
    static URL: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"(?i)^[a-z]+://").unwrap());
    static FILE: LazyLock<Regex> =
        LazyLock::new(|| Regex::new(r"^[A-Za-z0-9_.@-]+\.[A-Za-z0-9_-]+$").unwrap());
    !text.is_empty()
        && text.encode_utf16().count() <= 300
        && !text.contains(['\n', '<', '>', '{', '}'])
        && !URL.is_match(text)
        && (text.contains(['/', '\\']) || FILE.is_match(text))
}
fn paths(
    value: &Value,
    key: &str,
    depth: usize,
    output: &mut Vec<String>,
    seen: &mut HashSet<String>,
) {
    if depth > 4 {
        return;
    }
    match value {
        Value::String(text) if path_key(key) && file_path(text) => {
            let text = text.trim().to_owned();
            if seen.insert(text.clone()) {
                output.push(text);
            }
        }
        Value::Array(values) => {
            for value in values {
                paths(value, key, depth + 1, output, seen);
            }
        }
        Value::Object(values) => {
            for (key, value) in values {
                paths(value, key, depth + 1, output, seen);
            }
        }
        _ => {}
    }
}
fn operations(tool: &str, input: &Value) -> Vec<(String, String)> {
    let entries = input.as_array().or_else(|| input["content"].as_array());
    let patch = entries
        .into_iter()
        .flatten()
        .filter_map(|entry| {
            let kind = match entry["type"].as_str()? {
                "edit_file" | "update_file" | "move_file" => "edit",
                "write_file" => "write",
                "delete_file" => "delete",
                _ => return None,
            };
            let path = entry["path"]
                .as_str()
                .filter(|s| !s.is_empty())
                .or_else(|| entry["old_path"].as_str())?
                .trim();
            (!path.is_empty()).then(|| (kind.to_owned(), path.to_owned()))
        })
        .collect::<Vec<_>>();
    if !patch.is_empty() {
        return patch;
    }
    let Some(kind) = kind(tool) else {
        return Vec::new();
    };
    let mut collected = Vec::new();
    paths(input, "", 0, &mut collected, &mut HashSet::new());
    collected
        .into_iter()
        .map(|path| (kind.into(), path))
        .collect()
}

pub fn summarize(head: &SessionHead, messages: &[Message]) -> Vec<SessionFileActivity> {
    let mut activities = Vec::<SessionFileActivity>::new();
    let mut indexes = HashMap::new();
    for message in messages {
        for part in &message.parts {
            if let MessagePart::Tool {
                tool,
                state,
                time_created,
                ..
            } = part
                && let Some(input) = &state.input
            {
                for (kind, path) in operations(tool, input) {
                    let index = *indexes
                        .entry((kind.clone(), path.clone()))
                        .or_insert_with(|| {
                            activities.push(SessionFileActivity {
                                reference: head.reference.clone(),
                                project_identity_key: head.project_identity.key.clone(),
                                path,
                                kind,
                                count: 0,
                                latest_time: time_created.unwrap_or(message.time_created),
                            });
                            activities.len() - 1
                        });
                    let activity = &mut activities[index];
                    activity.count += 1;
                    activity.latest_time = activity
                        .latest_time
                        .max(time_created.unwrap_or(message.time_created));
                }
            }
        }
    }
    activities.sort_by(|a, b| {
        b.latest_time
            .total_cmp(&a.latest_time)
            .then_with(|| crate::locale::compare(&a.path, &b.path))
    });
    activities
}
